snuba/clickhouse/translators/snuba/mappers.py

```diff
@@ -44,14 +47,76 @@ def attempt_map(
             and expression.table_name == self.from_table_name
         ):
             return ColumnExpr(
-                alias=expression.alias,
+                alias=expression.alias
+                or qualified_column(self.from_col_name, self.from_table_name or ""),
                 table_name=self.to_table_name,
                 column_name=self.to_col_name,
             )
         else:
             return None
 
 
+@dataclass(frozen=True)
+class ColumnToLiteralMapper(ColumnMapper):
+    """
+    Maps a column name into a hardcoded literal preserving the alias.
+    """
+
+    from_table_name: Optional[str]
+    from_col_name: str
+    to_literal_value: OptionalScalarType
+
+    def attempt_map(
+        self,
+        expression: ColumnExpr,
+        children_translator: SnubaClickhouseStrictTranslator,
+    ) -> Optional[LiteralExpr]:
+        if (
+            expression.table_name == self.from_table_name
+            and expression.column_name == self.from_col_name
+        ):
+            return LiteralExpr(
+                alias=expression.alias
+                or qualified_column(self.from_col_name, self.from_table_name or ""),
+                value=self.to_literal_value,
+            )
+        else:
+            return None
+
+
+@dataclass(frozen=True)
+class ColumnToFunctionMapper(ColumnMapper):
+    """
+    Maps a column into a function expression that preserves the alias.
+
+    WARNING: The alias of the FunctionCall object passed to the
+    constructor as `to_function` will be ignored. It will be replaced
+    with the one present in the expression being translated.
+    Otherwise the resulting query could be incorrect.
```

Comment:
"What makes FunctionCall so special that it should be treated differently?"

The reason is that the parameters of a Column in the other classes are only two strings. while a FunctionCall is a nested structure, the client is not passing the function name only (see where this is used: https://github.com/getsentry/snuba/pull/985/files#diff-99e8c1b63173e867092f9d19e3bebdcbR39-R51).
The alternatives are:
1) asking the callsite to pass two parameters: function_name and a tuple of function parameters instead of the FunctionCall object. Now they cannot pass the alias anymore
2) build an entire separate class hierarchy of expression builders without aliases like FunctionBuilder, ParameterBuilder, etc.
I don't think the second one is worth doing, but the first approach could be reasonable.


"If there's a reason, should this at least validate that the alias is None in a __post_init__ to minimize confusion from misuse?" 

This is a good idea, will do.

Reply:
Adopted approach one: pass the function name and parameters directly instead of a FunctionCall object.